SteamKit2/SteamKit2/Steam/CMClient.cs

```diff
@@ -212,26 +212,35 @@ public void Connect( ServerRecord? cmServer = null )
                         return;
                     }
 
-                    connection = CreateConnection( record.ProtocolTypes & Configuration.ProtocolTypes );
-                    connection.NetMsgReceived += NetMsgReceived;
-                    connection.Connected += Connected;
-                    connection.Disconnected += Disconnected;
-                    connection.Connect( record.EndPoint, ( int )ConnectionTimeout.TotalMilliseconds );
+                    var newConnection = CreateConnection( record.ProtocolTypes & Configuration.ProtocolTypes );
+
+                    var connectionRelease = Interlocked.Exchange( ref connection, newConnection );
+                    DebugLog.Assert( connectionRelease == null, nameof( CMClient ), "Connection was set during a connect, did you call CMClient.Connect() on multiple threads?" );
+
+                    newConnection.NetMsgReceived += NetMsgReceived;
+                    newConnection.Connected += Connected;
+                    newConnection.Disconnected += Disconnected;
+                    newConnection.Connect( record.EndPoint, ( int )ConnectionTimeout.TotalMilliseconds );
                 }, TaskContinuationOptions.ExecuteSynchronously ).ContinueWith( t =>
-              {
+                {
                     if ( t.IsFaulted )
                     {
-                      LogDebug( nameof( CMClient ), "Unhandled exception when attempting to connect to Steam: {0}", t.Exception );
+                        LogDebug( nameof( CMClient ), "Unhandled exception when attempting to connect to Steam: {0}", t.Exception );
                         OnClientDisconnected( userInitiated: false );
                     }
+
+                    lock ( connectionLock )
```

Comment:
This will result in a deadlock with Disconnect, as it holds connectionLock to await this Task